### PR TITLE
removed restrictions for custom optimizer

### PR DESCRIPTION
**Pull request by CalogeroZarbo**
Removed the restriction to be able to use any optimizer of choice. A warning message has been added.

### CODE REVIEW

**Comment by tjruwase** [comment]
Looks good

**Comment by ShadenSmith** [comment]
Thanks for your contribution to DeepSpeed!

**Comment by CalogeroZarbo** [comment]
Thank you too for this amazing lib!

**Comment by ShadenSmith** [comment]
I was thinking this over some more and I'm worried that the lone warning is easy to miss in a typically-long logfile. That could lead to some "silent" divergence errors.

What if we keep the warning, but require an additional config in the JSON such as `"zero_allow_untested_optimizer" : true` in order to "unlock" the untested optimizer? If the flag is not set, we raise an error and provide instructions and the appropriate warnings.

**Comment by CalogeroZarbo** [comment]
Yep @ShadenSmith that seems to me as a very good approach to avoid problems difficult do debug in a complex training system. I'll fork again and make another pull-request if that's ok for you. Let me know.

Cheers,
Cal

**Comment by tjruwase** [comment]
@CalogeroZarbo this sounds good. Thanks!

**Comment by ShadenSmith** [comment]
Perfect, thanks @CalogeroZarbo !

**Comment by FWkey** [comment]
what are the possible errors if i use custom optimizer? or how to ensure its correctness, should i do some tests?

**Comment by tjruwase** [comment]
@FWkey, this is a tricky question. You could compare training loss for a smaller model of pytorch DDP vs ZeRO
